include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp

```diff
@@ -220,56 +224,116 @@ struct DeviceReduceBlockWise : public DeviceReduce<InElementwiseOperation, AccEl
-                                                                       InElementwiseOperation,
-                                                                       AccElementwiseOperation,
-                                                                       PropagateNan,
-                                                                       BetaIsZero,
-                                                                       BlockSize,
-                                                                       MThreadClusterSize,
-                                                                       KThreadClusterSize,
-                                                                       MThreadSliceSize,
-                                                                       KThreadSliceSize,
-                                                                       InSrcVectorDim,
-                                                                       InSrcVectorSize,
-                                                                       OutDstVectorSize>;
-
             float avg_time = 0;
 
-            const auto kernel = kernel_reduce_blockwise<GridwiseReduce,
-                                                        NeedIndices,
-                                                        InDataType,
-                                                        OutDataType,
-                                                        AccDataType,
-                                                        IndexDataType,
-                                                        InGridDesc_M_K,
-                                                        OutGridDesc_M,
-                                                        InElementwiseOperation,
-                                                        AccElementwiseOperation>;
-
-            avg_time = launch_and_time_kernel(stream_config,
-                                              kernel,
-                                              dim3(arg.gridSize),
-                                              dim3(BlockSize),
-                                              0,
-                                              in_grid_desc_m_k,
-                                              out_grid_desc_m,
-                                              arg.in_elementwise_op_,
-                                              arg.acc_elementwise_op_,
-                                              arg.alpha_,
-                                              arg.in_dev_,
-                                              arg.beta_,
-                                              arg.out_dev_,
-                                              nullptr,
-                                              arg.out_indices_dev_);
-
-            return (avg_time);
+            if constexpr(use_threadwise_reduce_kernel)
+            {
+                using GridwiseReduce =
+                    GridwiseReduction_mk_to_m_threadwise<InDataType,
```

Comment:
why we would want to use GridwiseReduction_mk_to_m_threadwise inside DeviceReduceBlockWise?

Reply:
> why we would want to use GridwiseReduction_mk_to_m_threadwise inside DeviceReduceBlockWise? why not put it inside DeviceReduceThreadWise?
> 
> This would be cleaner:
> 
> 1. DeviceReduceThreadwise use GridwiseReduction_mk_to_m_threadwise

This is just for using less DeviceOps,  no other reason,  the two can be easily combined in DeviceOp

> 2. Combine DeviceReduceMultiblockAtomicAdd and DeviceReduceBlockwise into DeviceReduceMultiBlockwise and use GridwiseReduction_mk_to_m_multiblock underneath

As we have discussed, if combining `DeviceReduceMultiblockAtomicAdd` and `DeviceReduceBlockwise` into one, the DeviceOp need to  choose a deterministic value of `blockGroupSize` in runtime,  but it is not easy to get a best choice.  It is very possible that `blockGroupSize > 1` is chosen, but `blockGroupSize == 1` can actually give better performance and lose the chance to be used.  

Another reason is that even when the two are being combined into one DeviceOp,  there is still big difference between the two executing paths in DeviceOp, like the follow
    if (blockGroupSize > 1) {
         // Using InMemoryDataOperationEnum::AtomicAdd  as kernel template parameter OutMemoryDataOperation
        // Clean the out_buffer to zero 
    }
    else {
        // Using InMemoryDataOperationEnum::Add as kernel template parameter OutMemoryDataOperation
  }

  Both the two executing paths and the kernels need be compiled as they are discern through runtime variant `blockGroupSize`.

Reply:
Changed!